Automattic/wp-calypso
# Generate variations from variation types

This PR was a bit of a challenge because redux is new to me and took a bit of time to wrap my head around.

Variations are also tricky so I had to work-out the logic for generating new possible variations based on provided variation types while also taking into account existing variations (either from state or based on an API call). Basically implementing something similar to [this](https://github.com/woocommerce/woocommerce/blob/e4a6882eaf878428d1f098f8755cd915373f258a/includes/class-wc-ajax.php#L583) and [this](https://github.com/woocommerce/woocommerce/blob/71d0702523bcc365403dd3b84b945da5218d88b4/includes/wc-core-functions.php#L1051) in WooCommerce core.

I don't expect this PR to be perfect, but I think it's time to get some eyes/feedback on it so I can make some improvements. In terms of UI or front-facing changes, it'll generate rows in the variation table which will later house our settings like price (future PR). [Preview](https://cloudup.com/csgiqd9soGP).

So this PR:

* Hooks up the variation types form to variation generation so new variations can be created as new types are added/updated and new variations are made possible.
* Introduces state so that the types form can generate and pass variation data to the rest of the form.

To Test:

* Visit `products/add`
* Add a type named 'Color' with a 'Red' value.
* Add a type named 'Size' with a 'Small' value.
* Press "Create Variations".
* One variation should be created titled "Red - Small".
* Add a 'Blue' color.
* Press "Update Variations".
* Another variation should be created, titled "Blue - Small".
* Add a 'Medium' size.
* Press "Update Variations".
* Two new variations should be created, title "Red - Medium", and "Blue - Medium".
* Feel free to add additional types, try different labels/values, etc.

If needed, state can be cleared with: `localStorage.clear(); indexedDB.deleteDatabase( 'calypso' );`. The ability to remove variations will come in a later PR.
